source/extensions/filters/network/thrift_proxy/header_transport_impl.cc

```diff
@@ -146,11 +147,18 @@ bool HeaderTransportImpl::decodeFrameStart(Buffer::Instance& buffer, MessageMeta
       throw EnvoyException(absl::StrCat("invalid header transport header count ", num_headers));
     }
 
+    Envoy::Http::StatefulHeaderKeyFormatterOptRef formatter = metadata.headers().formatter();
     while (num_headers-- > 0) {
       std::string key_string = drainVarString(buffer, header_size, "header key");
       // LowerCaseString doesn't allow '\0', '\n', and '\r'.
       key_string =
           absl::StrReplaceAll(key_string, {{std::string(1, '\0'), ""}, {"\n", ""}, {"\r", ""}});
+
+      if (formatter) {
+        // TODO(rgs1): do we also want to preserve the removed characters from above?
```

Comment:
yes, the removal is also a violation of the spec, see this commit on the clarification of the spec: https://github.com/apache/thrift/commit/e165fa3c85d00cb984f4d9635ed60909a1266ce1

this should be moved to be before the `StrReplaceAll` call.

Reply:
cc: @wbpcode

Reply:
I agree that we should move the formatter code before StrReplaceAll call.

Reply:
I guess we're waiting on this, but otherwise seems good to me.

Reply:
to clarify I'm totally ok if we merge the current version first, cut a release, and address this issue in a follow up PR later.

although they are both spec violations, case sensitivity is far more severe/common than those special characters. fixing wrong case sensitivity will fix probably 99% of the spec violations. so if fixing this extra case is too complicated it might worth it to get the other fix out first. (but I'll leave that to y'all to decide)